// @smart-webcomponents-angular/scheduler/esm2020/scheduler/smart.scheduler.module.mjs
import { NgModule } from '@angular/core';
import { SchedulerComponent } from './smart.scheduler';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import * as i0 from "@angular/core";
export class SchedulerModule {
}
SchedulerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: SchedulerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SchedulerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: SchedulerModule, declarations: [SchedulerComponent], exports: [SchedulerComponent] });
SchedulerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: SchedulerModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.6", ngImport: i0, type: SchedulerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [SchedulerComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [SchedulerComponent]
                }]
        }] });
